pkg/ebpf/bpf_error.go

```diff
@@ -0,0 +1,159 @@
+	BPF_ERR_UNSPEC BPFErrorType = iota
+
+	BPF_ERR_MAP_UPDATE_ELEM
+	BPF_ERR_MAP_DELETE_ELEM
+	BPF_ERR_GET_CURRENT_COMM
+	BPF_ERR_TAIL_CALL
+)
+
+var stringMap = map[BPFErrorType]string{
+	BPF_ERR_UNSPEC: "BPF_ERR_UNSPEC",
+
+	BPF_ERR_MAP_UPDATE_ELEM:  "BPF_ERR_MAP_UPDATE_ELEM",
+	BPF_ERR_MAP_DELETE_ELEM:  "BPF_ERR_MAP_DELETE_ELEM",
+	BPF_ERR_GET_CURRENT_COMM: "BPF_ERR_GET_CURRENT_COMM",
+	BPF_ERR_TAIL_CALL:        "BPF_ERR_TAIL_CALL",
+}
+
+var errorMap = map[BPFErrorType]string{
+	BPF_ERR_UNSPEC: "Unspecifed BPF error",
+
+	BPF_ERR_MAP_UPDATE_ELEM:  "Failed to add or update a map element",
+	BPF_ERR_MAP_DELETE_ELEM:  "Failed to delete a map element",
+	BPF_ERR_GET_CURRENT_COMM: "Failed to get current command",
+	BPF_ERR_TAIL_CALL:        "Failed to tail call",
+}
+
+func (b BPFErrorType) String() string {
+	if bpfErr, found := stringMap[b]; found {
+		return bpfErr
+	}
+
+	return stringMap[BPF_ERR_UNSPEC]
+}
+
+type BPFError struct {
+	ret      uint64
+	id       uint32
+	line     uint32
+	function [BPF_MAX_ERR_FUNC_LEN]byte
+}
+
+func (b BPFError) Error() string {
+	if bpfErr, found := errorMap[BPFErrorType(b.id)]; found {
+		return bpfErr
+	}
+
+	return errorMap[BPF_ERR_UNSPEC]
+}
+
+func (b BPFError) Return() int64 {
+	return int64(b.ret)
+}
+
+func (b BPFError) ID() uint32 {
+	return uint32(b.id)
+}
+
+func (b BPFError) Type() BPFErrorType {
+	return BPFErrorType(b.id)
+}
+
+func (b BPFError) Line() uint32 {
+	return b.line
+}
+
+func (b BPFError) Function() []byte {
+	return b.function[:]
+}
+
+func (b BPFError) FunctionAsString() string {
+	errFunc := string(b.function[:])
+	errFuncEnd := strings.Index(errFunc, "\x00")
+
+	return errFunc[:errFuncEnd]
+}
+
+func (b BPFError) returnSize() int {
+	return int(unsafe.Sizeof(b.ret))
+}
+
+func (b BPFError) idSize() int {
+	return int(unsafe.Sizeof(b.id))
+}
+
+func (b BPFError) lineSize() int {
+	return int(unsafe.Sizeof(b.line))
+}
+
+func (b BPFError) functionSize() int {
+	return int(len(b.function))
+}
+
+func (b BPFError) Size() int {
+	return b.returnSize() + b.idSize() + b.lineSize() + b.functionSize()
+}
+
+func (b *BPFError) Decode(rawBuffer []byte) error {
+	if len(rawBuffer) < b.Size() {
+		return fmt.Errorf("can't decode error raw data - buffer of %d should be of %d bytes", len(rawBuffer), b.Size())
+	}
+
+	b.ret = binary.LittleEndian.Uint64(rawBuffer[0:8])
+	b.id = binary.LittleEndian.Uint32(rawBuffer[8:12])
+	b.line = binary.LittleEndian.Uint32(rawBuffer[12:16])
+	copy(b.function[:], rawBuffer[16:56])
+
+	return nil
+}
+
+func (t *Tracee) processBPFErrors() {
+	for {
+		select {
+		case rawData := <-t.bpfErrorsChannel:
+			if len(rawData) == 0 {
+				continue
+			}
+
+			bpfErr := BPFError{}
+			if err := bpfErr.Decode(rawData); err != nil {
+				t.handleError(err)
+				continue
+			}
+
+			logger.Error(bpfErr.Error(),
+				"id", bpfErr.ID(),
+				"type", bpfErr.Type().String(),
+				"ret", bpfErr.Return(),
+				"func", bpfErr.FunctionAsString(),
+				"line", bpfErr.Line(),
+			)
```

Comment:
@NDStrahilevitz @yanivagman I suppose we don't want bpf error being printing to the standard output, right? Unless any occurrence of it is a case of panic. Should we logger.Debug() here or create another logger to output through Error() to stderr (since the pkg one default is to stdout)?

Reply:
Ignore that. I assume that the user must redirect the error to not taint the proper output.